File: frontend/src/api/attendance.ts
import api from '@/api/client'

export interface Attendance {
  id: number
  employee_id: number
  employee_name?: string
  date: string  // YYYY-MM-DD
  check_in_time?: string  // HH:MM:SS
  check_out_time?: string  // HH:MM:SS
  status: 'present' | 'late' | 'absent' | 'sick'
  notes?: string
  created_at: string
  updated_at: string
}

export interface AttendanceCreate {
  employee_id: number
  date: string  // YYYY-MM-DD
  check_in_time?: string  // HH:MM:SS
  check_out_time?: string  // HH:MM:SS
  status: 'present' | 'late' | 'absent' | 'sick'
  notes?: string
}

export const attendanceApi = {
  getAll: async (params?: {
    employee_id?: number
    from?: string  // YYYY-MM-DD
    to?: string    // YYYY-MM-DD
    status?: string
  }) => {
    const queryParams = new URLSearchParams()
    if (params?.employee_id) queryParams.append('employee_id', params.employee_id.toString())
    if (params?.from) queryParams.append('from', params.from)
    if (params?.to) queryParams.append('to', params.to)
    if (params?.status) queryParams.append('status', params.status)

    const response = await api.get(`/attendances?${queryParams.toString()}`)
    return response.data as Attendance[]
  },

  getById: async (id: number) => {
    const response = await api.get(`/attendances/${id}`)
    return response.data as Attendance
  },

  create: async (data: AttendanceCreate) => {
    const response = await api.post('/attendances', data)
    return response.data as Attendance
  },

  update: async (id: number, data: Partial<AttendanceCreate>) => {
    const response = await api.put(`/attendances/${id}`, data)
    return response.data as Attendance
  },

  delete: async (id: number) => {
    await api.delete(`/attendances/${id}`)
  },
}

File: backend/app/schemas/attendance.py
"""
Pydantic schemas for Attendance management
"""
from pydantic import BaseModel, Field, validator
from typing import Optional
from datetime import datetime, date, time


class AttendanceBase(BaseModel):
    """Base attendance schema"""
    employee_id: int = Field(..., gt=0)
    date: date
    check_in_time: Optional[time] = None
    check_out_time: Optional[time] = None
    status: str = Field(..., description="present, late, absent, sick")
    notes: Optional[str] = Field(None, max_length=500)
    
    @validator('status')
    def validate_status(cls, v):
        """Validate attendance status"""
        valid_statuses = ['present', 'late', 'absent', 'sick']
        if v not in valid_statuses:
            raise ValueError(f'Status must be one of: {", ".join(valid_statuses)}')
        return v


class AttendanceCreate(AttendanceBase):
    """Schema for creating attendance"""
    pass


class AttendanceUpdate(BaseModel):
    """Schema for updating attendance"""
    check_in_time: Optional[time] = None
    check_out_time: Optional[time] = None
    status: Optional[str] = None
    notes: Optional[str] = Field(None, max_length=500)
    
    @validator('status')
    def validate_status(cls, v):
        """Validate attendance status"""
        if v is not None:
            valid_statuses = ['present', 'late', 'absent', 'sick']
            if v not in valid_statuses:
                raise ValueError(f'Status must be one of: {", ".join(valid_statuses)}')
        return v


class AttendanceResponse(AttendanceBase):
    """Schema for attendance response"""
    id: int
    created_at: datetime
    updated_at: datetime
    
    # Include employee details
    employee_name: Optional[str] = None
    
    class Config:
        from_attributes = True
        json_encoders = {
            datetime: lambda v: v.isoformat(),
            date: lambda v: v.isoformat(),
            time: lambda v: v.isoformat()
        }
